Introduce special DeclNames for subscripts

ahoppen opened the pull request:
This is a preparation to use special DeclNames that don't carry identifiers e.g. for subscripts, initializers, and destructors which shall at some point fix [SR-2575](https://bugs.swift.org/browse/SR-2575) as outlined in #5537 and #6305.

ahoppen commented:
@jrose-apple: I've taken a more conservative approach now. Could you maybe have a quick look if this is going in the right direction? It's still kind of WIP and I haven't created the matching PR to the LLDB repo yet, so no need for running the test or doing an in-depth review. Thanks!

ahoppen commented:
I'm making fairly good progress on this PR now and just have a few issues to fix. For that, I've got a quick question:

Subscripts are currently represented in SIL as `<Type>.subscript!(getter|setter)` which is the same spelling as for an instance variable named `subscript`. I.e. the accessor path for the following two is the same:

```
class Foo {
  subscript() -> String {
    return ""
  }
  
  var `subscript`: String
}
```

namely `#Foo.subscript!getter.1`

Until now this wasn't an issue since both of them were represented by the identifier `Id_subscript` so name lookup found either. Furthermore, the above code doesn't compile right now and complains about a redeclaration of `subscript`. However, with the introduction of special `DeclName`s, the above code is actually valid. So we would need a way to disambiguate the two in SIL to get the same expressive power there. My proposal is to rename the actual subscript reference in SIL to something like `<Type>.[]!getter`, `[]` being a constant string (nothing is meant to go in between the brackets), but I'm open to other solutions as well. The only requirement for the representation is that it isn't a valid identifier on the surface language.

Do you think this syntax would fit in the current spirit of SIL (which I really don't know a lot about) or is there another syntax coming to your mind that would be better? Thanks!

jrose-apple commented:
Sorry for continuing to ignore this series, even after Branch Day—lots of 3.1 bugs to keep our hands full!

I'm no SIL expert, but I think expanding the set of allowed characters seems a bit iffy. Don't we have the same problem with `init` and `deinit` stepping on regular declarations? Maybe we should just add a leading '!' or something, i.e. `!init` and `!subscript!getter`. @jckarter, @swiftix, any thoughts on this particular issue?

jckarter commented:
Or \`subscript\`!1 ?

ahoppen commented:
Subscript in backticks sounds good. While implementing it, I found two issues that prevents generated SIL files from being parsed again. I just wanted to point to them (workarounds would be fine for me right now) since they make further development nearly impossible: [SR-3773](https://bugs.swift.org/browse/SR-3773), [SR-3774](https://bugs.swift.org/browse/SR-3774)

ahoppen committed:
Add test cases for special subscript name

ahoppen commented:
Please test with following pull request:
apple/swift-lldb#106

@swift-ci Please test

ahoppen commented:
Please test with following pull request:
apple/swift-lldb#106

@swift-ci Please test

ahoppen commented:
Hitting this failure: https://lists.swift.org/pipermail/swift-dev/Week-of-Mon-20170130/003954.html
Hoping a rerun solves the issue.

Please test with following pull request:
apple/swift-lldb#106

@swift-ci Please test

ahoppen commented:
@jrose-apple: The PR is done so far from my side, could you (or someone else) review the changes?

jrose-apple commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -243,10 +243,13 @@ static void printValueDecl(ValueDecl *Decl, raw_ostream &OS) {
+  if (Decl->isOperator()) {
+    OS << '"' << Decl->getBaseName() << '"';
+  } else if (Decl->getBaseName() == Decl->getASTContext().Id_subscript) {
+    OS << '`' << Decl->getBaseName() << '`';
```
Also, it seems like for SIL it might make sense to have one consistent escaping syntax. Quotes might be fine. @swiftix, @gottesmm?

jrose-apple replied:
Still need clarity on this.

ahoppen replied:
See my comment on `swift-api-digester.cpp:1029` further down.

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2387,13 +2387,13 @@ bool FailureDiagnosis::diagnoseGeneralMemberFailure(Constraint *constraint) {
     if (!nameStr.getAsInteger(10, Value) && Value < tuple->getNumElements()) {
       fieldIdx = Value;
-    } else {
+    } else if (!memberName.isSpecialName()) {
```
Is this check necessary? It seems like a special name just wouldn't match anyway.

ahoppen replied:
This keeps the call to `getBaseName` further down from being executed which would crash in case of a special `DeclName`

jrose-apple commented on lib/Sema/CSGen.cpp:
```diff
@@ -1168,7 +1166,7 @@ namespace {
                                      CurDC);
       } else {
-        CS.addValueMemberConstraint(baseTy, Context.Id_subscript,
+        CS.addValueMemberConstraint(baseTy, DeclName::createSubscript(),
```
Does this correctly distinguish between `subscript` and `subscript()`?

ahoppen replied:
I think so. I don't understand your specific concern though…

jrose-apple replied:
It's the difference between a reference to "any subscript" and "a subscript with zero arguments", or between a base DeclName and a compound DeclName. Just wanted to make sure this was producing a base DeclName and not a compound one.

jrose-apple commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2795,8 +2795,16 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
     if (baseObjTy->isAnyObject() && argumentLabels) {
-      memberName = DeclName(TC.Context, memberName.getBaseName(),
-                            argumentLabels->Labels);
+      switch (memberName.getKind()) {
```
Maybe there should be a general `withArgumentLabels` operation on DeclName?

jrose-apple commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -5174,22 +5174,30 @@ static Optional<unsigned> scorePotentiallyMatchingNames(DeclName lhs,
+    }
+  } else {
+    if (lhs.getKind() == rhs.getKind()) {
+      score = 0;
```
What about argument labels?

ahoppen replied:
Aren't they handled further down (line 5205 - 5210)?

jrose-apple replied:
Oh, whoops. I thought this was an early return for some reason.

jrose-apple commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -297,22 +298,36 @@ namespace {
 public:
-  using internal_key_type = StringRef;
-  using external_key_type = Identifier;
+  using internal_key_type = std::pair<DeclNameKind, StringRef>;
```
This is good, use something like this for the ClangImporter lookup tables, please.

jrose-apple commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -1770,12 +1770,34 @@ Identifier ModuleFile::getIdentifier(IdentifierID IID) {
+    // Subscript
+    return DeclName::createSubscript();
+  } else {
+    // Serialized prior to special DeclNames. Just read the normal identifier
```
At this point the module format isn't stable, so we can just make a breaking change here. It seems more important to look forward than backward; in a forward world, we've introduced a new special kind that *this* compiler treats as a normal name. We could either import it as a normal name or fail.

ahoppen replied:
OK, I didn't want to be the first one to increase the major version though ;-) But with your OK, I'll go ahead ;-)

jrose-apple replied:
We're treating the minor version as a breaking version for now, please continue doing that.

jrose-apple reviewed (commented):
Thank you for continuing to push this forward, Alex. I find I'm still really uncomfortable with this change because it erases the distinction between operations that take compound names and those that only take base names (whether normal or special). I'm going to go back to saying that we need to change Identifier, not DeclName. @DougGregor?

ahoppen commented on include/swift/AST/Decl.h:
```diff
@@ -4698,6 +4700,8 @@ class AbstractFunctionDecl : public ValueDecl, public DeclContext {
   GenericEnvironment *getLazyGenericEnvironmentSlow() const;
 
 public:
+  Identifier getName() const { return getFullName().getBaseName(); }
```
OK, I wanted to keep the code change as small as possible and not litter the PR with renames from `getName` to `getBaseName`.

ahoppen commented on include/swift/AST/Identifier.h:
```diff
@@ -395,6 +445,16 @@ class DeclName {
+  StringRef str() const {
+    switch (getKind()) {
+    case DeclNameKind::Normal:
+      return getBaseName().str();
```
Renaming sounds like the way to go since IIRC this behaviour is needed in quite a few places.

ahoppen commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1733,7 +1734,7 @@ void ASTMangler::appendProtocolConformance(const ProtocolConformance *conformanc
               fileUnit->getDiscriminatorForPrivateValue(behaviorStorage).str());
     appendProtocolName(conformance->getProtocol());
-    appendIdentifier(behaviorStorage->getName().str());
+    appendIdentifier(behaviorStorage->getBaseName().getBaseName().str());
```
This basically asserts that the base name is not special because then the second `getBaseName()` will crash. Maybe an `assert` would indeed be cleaner.

ahoppen commented on lib/AST/Decl.cpp:
```diff
@@ -190,7 +190,7 @@ DescriptiveDeclKind Decl::getDescriptiveKind() const {
      }
 
-     if (!func->getName().empty() && func->getName().isOperator())
+     if (!func->getName().empty() && func->isOperator())
```
It doesn't right now but it seems like a reasonable thing to add there.

ahoppen commented on lib/AST/Decl.cpp:
```diff
@@ -4175,10 +4174,12 @@ DeclName AbstractFunctionDecl::getEffectiveFullName() const {
-                                      ArrayRef<Identifier>());
-
+          if (afd->getFullName().getKind() == DeclNameKind::Subscript) {
+            return DeclName::createSubscript(ctx, ArrayRef<Identifier>());
```
I don't know why I changed this in the first place. I'll change it back to how it was

ahoppen commented on lib/AST/Identifier.cpp:
```diff
@@ -79,9 +96,29 @@ int Identifier::compare(Identifier other) const {
+  return Identifier(SubscriptIdentifierData);
+}
+
+DeclNameKind DeclName::getKind() const {
```
I think the problem was that if you initialise `SubscriptIdentifierData ` as a private static member on `DeclName` in the header of `DeclName` it gets a different value for each library that import `Identifier.h`. This worked around the issue.

jrose-apple replied:
Sorry, `static` as a member doesn't mean the same thing as `static` elsewhere. As long as it's defined out of line there's no problem declaring it ahead of time.

ahoppen commented on lib/AST/NameLookup.cpp:
```diff
@@ -883,9 +883,9 @@ UnqualifiedLookup::UnqualifiedLookup(DeclName Name, DeclContext *DC,
   }
 
-  // TODO: Does the debugger client care about compound names?
+  // TODO: What about special DeclNames in the debugger?
```
Whoops, deleted the wrong todo, the other one should have stayed

ahoppen commented on lib/AST/Mangle.cpp:
```diff
@@ -103,6 +103,13 @@ void Mangler::mangleIdentifier(Identifier ident, OperatorFixity fixity) {
 }
 
+/// Mangle a DeclName into the buffer.
+void Mangler::mangleDeclName(DeclName name, OperatorFixity fixity) {
```
I'll look into it if we can get rid of this method.

ahoppen commented on lib/ClangImporter/SwiftLookupTable.cpp:
```diff
@@ -1157,27 +1179,50 @@ void SwiftLookupTableWriter::writeExtensionContents(
-      return key;
+      switch (key.first) {
+      case swift::DeclNameKind::Normal:
+        return DeclName(ctx.getIdentifier(key.second));
```
Ah, I didn't think about a growing identifier table. Now it makes sense that `StringRef` was used throughout that part. I'll try and change some stuff back to strings then.

ahoppen commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2602,7 +2602,8 @@ namespace {
           if (errorWrapper) {
             auto enumeratorValue = cast<ValueDecl>(enumeratorDecl);
-            auto alias = importEnumCaseAlias(enumeratorValue->getName(),
+            auto name = enumeratorValue->getFullName().getBaseName();
```
Oh, yeah, good catch ;-)

ahoppen commented:
Thanks for the review. It explained a few things I didn't quite understand before. 😄
I'll wait with any further changes until we've reached a decision whether this is the way forward or not. If we decide to go a more `Identifier` centric approach, there is still my initial idea of using `$subscript` similar to `$match` in `switch` statements, which we could revive again.

ahoppen commented:
Just a friendly ping that it would be great if you could discuss internally if we want to continue with the `DeclName` based approach or try to model the special names using `Identifier`. I really don't feel like I'm able to judge what's best in the overall picture.

rjmccall commented:
It sounds to me like the base name of a DeclName should use a new DeclBaseName type that's either an identifier or one of these special names.  We shouldn't contaminate the entire Identifier type with the possibility of special names, but we also shouldn't pretend that special names only make sense at the DeclName level.

ahoppen commented:
Sounds very reasonable to me.
And I guess then change some of the places where we currently (prior to this PR) use `Identifier`s to use the new `DeclBaseName` if it could be a special name.

rjmccall commented:
Yeah.  That might not be a very fun project, but it's probably the right thing to do.  I'd wait for @jrose-apple's blessing before getting started, though.

jrose-apple commented:
I'm sad to have *three* types, because I'm not sure Identifier will be pulling its weight after this, but it's always easier to collapse back down to two if we change our minds later, and right now we really have identified three different kinds of entities.

jrose-apple commented:
Thanks for weighing in, @rjmccall.

ahoppen commented:
Cool, I'll try and tackle this one more time then ;-)

ahoppen committed:
Introduce special decl names

Special DeclNames represent names that do not have an identifier in the
surface language. This implies serializing the information about whether
a name is special together with its identifier (if it is not special)
in both the module file and the swift lookup table.

ValueDecls are not guaranteed to have a name anymore and thus getNameStr
is moved from ValueDecl to subclasses that guarantee to be backed by
identifiers. Other calls need to handle special names if the context
doesn't guarantee that the declaration's name is non-special.

ahoppen committed:
[Gardening] Change calls of getName().isOperator() to isOperator()

ahoppen committed:
Change the backing base name type of DeclName to DeclBaseName

The new DeclBaseName type will later abstract over normal Identifiers as
they exist now and special names that don't have an identifier in the
surface language like subscripts

ahoppen committed:
Preparations for special names

Special names will in the future represent names that don't have a
surface language identifier like subscript.

For this, change the backing type of DeclName to a new type DeclBaseName
that will later be able to represent special names.

ValueDecls may therefore no longer have an identifier name. Thus we move
the getName method from ValueDecl to the subtypes that are guaranteed
to have identifier names

ahoppen commented:
Please test with following pull request:
apple/swift-lldb#165

@swift-ci Please test

ahoppen commented:
OK, so I've implemented the changes with a new type `DeclBaseName` now. I personally feel like this could be a good solution, the impact of the changes seems to be far lower than with the other approaches. So, a review would be very welcome.

I've got one more open point which is that subscripts are currently mangled as something like `_TFC4main3Foog9subscriptFT_Si`.  I feel like we shouldn't include the string 'subscript' in the mangled name anymore, but that's something to be tackled in a subsequent PR, I'd say.

jrose-apple commented on include/swift/AST/Identifier.h:
```diff
@@ -203,7 +205,144 @@ namespace llvm {
+    return userFacingStr().compare(other.userFacingStr());
+  }
+
+  bool operator==(StringRef Str) const {
```
These seem too easy to misuse. Maybe we should go the Clang route: {{bool isStr(StringRef str)}}.

ahoppen replied:
What kind of misuse do you have in mind? I quite liked the brevity of `==` in `if`s. But I can, of course, change it

jrose-apple replied:
I guess I was worried about {{== "subscript"}} (and whatever else we add), and maybe that's being too cautious.

jrose-apple commented on include/swift/AST/Identifier.h:
```diff
@@ -203,7 +205,144 @@ namespace llvm {
+
+raw_ostream &operator<<(raw_ostream &OS, swift::DeclBaseName D);
+
+// Identifiers hash just like pointers.
```
Nitpick: comment.

jrose-apple commented on include/swift/AST/Identifier.h:
```diff
@@ -203,7 +205,144 @@ namespace llvm {
+// Identifiers hash just like pointers.
+template<> struct DenseMapInfo<swift::DeclBaseName> {
+  static swift::DeclBaseName getEmptyKey() {
+    return swift::DeclBaseName::getEmptyKey();
```
Nitpick: Why the extra level of indirection? It's not like `getEmptyKey` is going to be used anywhere else.

(I see that Identifier did this too, but it's weird there too.)

ahoppen replied:
I was just copying what `Identifier` did ;-). I'll change it.

jrose-apple commented on include/swift/AST/Mangle.h:
```diff
@@ -154,6 +154,8 @@ class Mangler {
   void mangleProtocolList(ArrayRef<Type> protocols);
   void mangleIdentifier(Identifier ident,
                         OperatorFixity fixity = OperatorFixity::NotOperator);
+  void mangleIdentifier(DeclBaseName ident,
```
What do we want to do about things like this? Is continuing to say `mangleIdentifier` the right way forward? I think it's fine for this patch (minimizes an already large changeset), but what do you think for follow-up?

ahoppen replied:
I would imagine that we will at some point change the mangling name of subscripts to something that doesn't contain the string "subscript" (I'd use a new special character in the mangling scheme). Then the special names would be handled on the call side of this function and we could remove it again.

jrose-apple replied:
Ah, right, we could do it based on the DeclKind. Okay, that seems reasonable.

jrose-apple commented on include/swift/Serialization/ModuleFormat.h:
```diff
@@ -344,6 +344,18 @@ enum SpecialModuleID : uint8_t {
+// have an identifier. As above, these IDs must \em not be renumbered or
+// reordered without incrementing VERSION_MAJOR.
+enum SpecialNameID : uint8_t {
+  SUBSCRIPT_ID = NUM_SPECIAL_MODULES,
```
This seems dangerous. How do people know whether to check NUM_SPECIAL_MODULES or NUM_SPECIAL_IDS? I think we should either fold the two lists together or allow the same value to represent multiple things (builtin module OR `subscript`, current module OR `init`, etc).

ahoppen replied:
Yeah, I think folding the two enums together would be the right way to go. Didn't really feel too comfortable with it but wanted an opinion on whether using the same enum for both special module names and other special names would be alright.

jrose-apple commented on lib/AST/ASTMangler.cpp:
```diff
@@ -390,7 +391,15 @@ void ASTMangler::appendDeclName(const ValueDecl *decl) {
     }
   } else {
-    appendIdentifier(decl->getName().str());
+    // FIXME: Should a mangled subscript name contain the string "substring"?
```
"subscript" :-) I tend to agree, we might as well pick an abbreviation that isn't used. But we can do that later.

ahoppen replied:
Whoops. Also, see my comment to `mangleIdentifier`.

jrose-apple commented on lib/AST/Identifier.cpp:
```diff
@@ -20,13 +20,21 @@
 #include "llvm/Support/ConvertUTF.h"
 using namespace swift;
 
+void *DeclBaseName::SubscriptIdentifierData =
```
Casting this to `char *` to initialize an Identifier is technically illegal. Maybe DeclBaseName should just store a `void *` and avoid the whole mess?

(Also, it would be nice if it actually *were* a string, so that you could get something useful in the debugger. But we'd have to make sure that string was 4-byte aligned.)

ahoppen replied:
Is there a quick explanation of why the cast is illegal or somewhere I can look it up? Proves once again that I know less about C++ than I thought ;-)

I wouldn't really want to change away from using `Identifier` as the underlying storage type of `DeclBaseName`. Somehow it doesn't feel right to me and would imho lead to unnecessary conversions whenever `getIdentifier` is called on `DeclBaseName`. Or did you mean changing the `char *` in `Identifier` to `void *`.

I decided against using a constant string because of the following (I might be wrong): Say we set `SubscriptIdentifierData` to "specialSubscript". Then, if somebody later decided to use a constant string named "specialSubscript" somewhere completely different it may end up being stored at the same location in the binary. Hence if you create an `Identifier` with this constant name it would be a special name. (But as I'm writing this, I realise that the chances of that are pretty low 😉)

Also: why would the string need to be 4-byte aligned?

jrose-apple replied:
Hm. Maybe I take it back about going from `void *` to `char *`. I was getting mixed up with levels of pointer, more so with the self-address assignment. And the `char *` better not be used as a string anyway (since there's no null byte).

The pointer needs to be 4-byte-aligned because you've claimed (and we rely on) DeclBaseName having two "spare bits" in its opaque representation. Using the address of a `void *` should handle that. :-)

The "conversions" between `void *` and `Identifier` would get optimized away in release builds, so I wouldn't have considered that such a blocker. PointerUnion does this all the time.

ahoppen replied:
OK, so keep everything as it is?

jrose-apple replied:
Yeah, I guess so. Sorry!

jrose-apple commented on lib/AST/Mangle.cpp:
```diff
@@ -103,6 +103,22 @@ void Mangler::mangleIdentifier(Identifier ident, OperatorFixity fixity) {
+/// Mangle an identifier into the buffer.
+void Mangler::mangleIdentifier(DeclBaseName ident, OperatorFixity fixity) {
+  StringRef str;
+  // FIXME: Should a serialized subscript name contain the string "substring"?
```
"substring" again :-)

jrose-apple commented on lib/AST/PrettyStackTrace.cpp:
```diff
@@ -39,7 +39,7 @@ void swift::printDeclDescription(llvm::raw_ostream &out, const Decl *D,
   if (auto *named = dyn_cast<ValueDecl>(D)) {
     if (named->hasName()) {
-      out << '\'' << named->getName() << '\'';
+      out << '\'' << named->getBaseName() << '\'';
```
This comment stands. :-)

jrose-apple commented on lib/ClangImporter/SwiftLookupTable.h:
```diff
@@ -45,6 +45,102 @@ class TypedefNameDecl;
+      return (unsigned)Val.Kind;
+    }
+  }
+  static bool isEqual(SerializedSwiftName LHS, SerializedSwiftName RHS) {
```
I think it's fine to just use `==` for this.

ahoppen replied:
It has a separate implementation from `==` in order to call through to `DenseMapInfo<StringRef>::getHashValue` to get correct correct comparisons for the `StringRef` tombstone and empty key. I guess calling `==` doesn't work because it uses `StringRef`'s `==` operator and since both the `StringRef` tombstone and empty key have length 0 they would end up being equal, which is incorrect. Didn't check though.

jrose-apple replied:
Oh, ouch. I guess that makes sense.

jrose-apple commented on lib/ClangImporter/SwiftLookupTable.h:
```diff
@@ -153,7 +249,7 @@ const uint16_t SWIFT_LOOKUP_TABLE_VERSION_MAJOR = 1;
 ///
 /// When the format changes IN ANY WAY, this number should be incremented.
-const uint16_t SWIFT_LOOKUP_TABLE_VERSION_MINOR = 14; // Swift 2 names
+const uint16_t SWIFT_LOOKUP_TABLE_VERSION_MINOR = 15; // Special names
```
Honestly I'm not sure why we even have major and minor versions here. @milseman?

jrose-apple replied:
Specifically, unlike swiftmodules, this table format is never distributed, just embedded into pcm files.

ahoppen replied:
It's not my place to worry right now though, is it?

jrose-apple replied:
Fair enough, this shouldn't block the patch.

milseman replied:
No idea, @DougGregor ?

jrose-apple commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2617,7 +2618,7 @@ diagnoseTypeMemberOnInstanceLookup(Type baseObjTy,
         ValueDecl *decl0 = overloadedFn->getDecls()[0];
 
-        if (decl0->getName() == decl0->getASTContext().Id_MatchOperator) {
+        if (decl0->getBaseName() == decl0->getASTContext().Id_MatchOperator) {
```
Nitpick: Why `getBaseName` here but `getFullName` elsewhere?

ahoppen replied:
My scheme was actually `getBaseName` everywhere and `getFullName` just if comparing against `self`. The idea behind it was that I don't care about the operator's argument labels in this context. In the `self` case, I wouldn't want to match some variable that has a composite name (or can guarantee that `self` never has a composite name).

jrose-apple replied:
Hm, okay. I guess it's questionable whether we'd make `self` a special name, even though we clearly want to do so for things like `init`.

ahoppen replied:
I never wanted to make `self` a special name (neither do I think it should be one). The above was just an explanation on why I chose to compare to `Id_self` using `getFullName()` (thus taking the `==` operator of `DeclName)` and compare to any other identifier (e.g. `Id_MatchOperator`) using `getBaseName()` (taking the `==` operator of `DeclBaseName`).

jrose-apple commented on lib/Sema/TypeCheckNameLookup.cpp:
```diff
@@ -434,18 +434,26 @@ static unsigned getCallEditDistance(DeclName argName, DeclName paramName,
+  if (argName.getBaseName().getKind() != paramName.getBaseName().getKind()) {
     return UnreasonableCallEditDistance;
   }
+  if (argName.getBaseName().getKind() == DeclBaseName::Kind::Normal) {
```
Can you make this an early return? If the two kinds are the same and they aren't a "Normal" kind, we can just return 0.

ahoppen replied:
So basically flip the `if`? Sounds like a good idea.

jrose-apple commented on lib/Serialization/Serialization.cpp:
```diff
@@ -1819,9 +1835,18 @@ void Serializer::writeCrossReference(const Decl *D) {
+  IdentifierID iid;
+  switch (val->getBaseName().getKind()) {
+  case DeclBaseName::Kind::Normal:
+    iid = addIdentifierRef(val->getBaseName().getIdentifier());
```
Maybe addIdentifierRef should just take a DeclBaseName instead?

ahoppen replied:
I didn't do that because of naming ;-)

If it would take `DeclBaseName` it would probably need to be renamed to `addBaseName`. But then, no special name is ever added to the serialised identifiers. But, taking `DeclBaseName` would probably be the right thing to do in retrospect. Naming issue still stands though…

jrose-apple replied:
I'm fine with `addDeclBaseNameRef`. Just because the "reference" is a hardcoded value for known names doesn't change the general purpose.

jrose-apple commented on test/Sema/subscripting.swift:
```diff
@@ -0,0 +1,16 @@
+// RUN: %target-run-simple-swift | %FileCheck %s
```
Nitpick: execution tests don't belong in test/Sema/. Can you move this to test/Interpreter/ instead?

jrose-apple commented on test/decl/subscript/subscripting.swift:
```diff
@@ -310,3 +310,12 @@ struct S4 {
+  }
+}
+
+SR2575().subscript() // expected-error{{type 'SR2575' has no member 'subscript'}}
```
We should probably make a note / bug to come back and improve this specific case later.

ahoppen replied:
What do you wan to change here? Offer a fixit to change the call to `subscript` to a proper subscript?

jrose-apple replied:
Something like that. Maybe something that makes it clear it's looking for functions *named* subscript rather than actual subscripts.

ahoppen replied:
OK, I'll open a follow-up JIRA once the PR has been landed.

jrose-apple commented on test/expr/unary/selector/selector.swift:
```diff
@@ -116,7 +116,7 @@ func testParseErrors3(_ c1: C1) {
 func testParseErrors4() {
   // Subscripts
-  _ = #selector(C1.subscript) // expected-error{{type 'C1.Type' has no subscript members}}
+  _ = #selector(C1.subscript) // expected-error{{type 'C1' has no member 'subscript'}}
```
I think we *are* going to want to support syntax like this for getting the selector of a subscript getter or setter some day: `#selector(getter:C1.subscript)`. It doesn't currently work but that doesn't mean it *shouldn't* work.

ahoppen replied:
I'd say that `C1.subscript` will always be invalid unless you declare a ``func `subscript`() {}``. Otherwise we would get into ambiguity issues if you declare both a subscript and a function named `subscript`.

jrose-apple replied:
That's what the `getter:` is for. Or whatever the no-argument function disambiguation syntax turns out to be.

ahoppen replied:
Sorry, but then I don't get why we want to support syntax like this. `#selector(getter:…)` would then need to implement some special lookup logic to handle the subscript case, I'd say.

jrose-apple replied:
…and I don't quite understand your objection. How do you want to spell "the selector of the getter for this subscript"?

I suppose we haven't had a huge need for it yet, but it's not an unreasonable thing to ask for.

ahoppen replied:
I think my point is that, from my point of view, whether or not we want to support selectors for subscripts is nothing to worry about for now and that we will only need to worry about how to implement it when the time comes.
Or maybe let's ask the other way round: What would you want to be changed here?

jrose-apple replied:
I'm concerned about the *diagnostic.* It implies that the compiler is interpreting this syntax as a reference to a member named 'subscript', which would lead to a user thinking they've done something wrong, rather than this simply being a feature that's not implemented yet.

It's not a deal-breaker, but I consider us not supporting subscript selectors to just be a bug, and therefore the compiler shouldn't tell the user they've done something wrong.

ahoppen replied:
Ah, OK, now I see. For now I'd suggest changing the error message to `type 'C1' has no member named 'subscript'` (adding the word 'named'). I guess then it's correct since `C1` doesn't have a member *named* subscript but only a subscript-member. 

In a subsequent PR we could then maybe improve the diagnostics by looking if the type would have a subscript-member (which was probably meant) and point out that selectors for subscripts are not yet supported. Once we do support them, this would turn into a fixit for the `getter` or `setter` syntax.

jrose-apple commented on tools/swift-api-digester/swift-api-digester.cpp:
```diff
@@ -1019,12 +1019,30 @@ static bool isFunctionTypeNoEscape(Type Ty) {
+  case DeclBaseName::Kind::Subscript:
+    return "subscript";
+  case DeclBaseName::Kind::Normal:
+    return llvm::StringSwitch<StringRef>(name.getIdentifier().str())
```
Rather than a StringSwitch, we should just escape anything that's a keyword.

ahoppen replied:
I don't want to escape all keywords here. A function named `default` should still be output unescaped. I'm basically escaping a new class of keywords here, namely those that represent special names (in which "subscript" is currently the only member)

jrose-apple replied:
I don't think that makes sense. You wouldn't want to see `func func` in the output of this tool. @nkcsgexi, what do you think? (Sorry to pull you in mid-discusson.)

nkcsgexi replied:
I think for swift-api-digester per se, being consistent is more important than being precise. So this change is fine as long as we always keep subscript escaped; however, this may not help the tool itself much.

jrose-apple commented on validation-test/Sema/type_checker_crashers_fixed/rdar27017206.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: not --crash %target-swift-frontend %s -typecheck
+// RUN: not %target-swift-frontend %s -typecheck
 // REQUIRES: asserts
```
Please drop the "REQUIRES: asserts" as well. (We're not good about this and often clean them up in large batches, but it's good to get in the habit.)

jrose-apple reviewed (commented):
This does look a lot better. It seems obvious now that we really do have three different kinds of things. Thanks, Alex!

The patch is still pretty big; it would be nice to factor out some of the changes that don't depend on DeclBaseName already existing, so we can review those separately and/or get those in first. I scanned through the entire thing today but that doesn't mean I didn't miss something.

ahoppen commented:
Thanks for the review.

I guess the only way I could split this in some sensible way would be across my different commits. This would mean, we could split it in up to 5 pull requests:

- Change calls of getName().isOperator() to isOperator()
  - Really small patch
- Change the backing base name type of DeclName to DeclBaseName
  - Introduce `DeclBaseName` as an implementation detail of `DeclName` but don't expose it yet
  - Basically only affects `Identifier.h`
- Preparations for special names
  - Expose `DeclBaseName` to the project but without special names for the moment
  - This basically adds a bunch of todos in places where special names will be handled later
- Introduce special decl names
  - Actually introduce special names and remove the todos added in the previous commit
  - This and the above commit are both still fairly big (~500 lines) but I can't think of any sensible way to break them apart so that it still compiles in between
- Add test cases for special subscript name
  - Should probably be merged together with the previous commit

jrose-apple commented:
That seems like a reasonable way to split it up. If (3) and (4) don't split nicely at least we'd have gotten (1) and (2) out of the way first, so they're not intermingled with everything else.

ahoppen commented:
I'll create PRs one by one now, splitting (3) and (4) since the commits are already separate. We can always merge them again if we want to.

Part 1 -> #8015
